File: src/advent/days/day_1.clj
(ns advent.days.day-1
  (:require [advent.util :as util]
            [clojure.string :as str]))

(defn part-1 [numbers]
  (->>
    numbers
    (filter (fn [num]
              (contains? numbers (- 2020 num))))
    (reduce *)))

(defn part-2 [numbers]
  (->>
    (for [x numbers y numbers z numbers] (vector x y z))
    (filter (fn [group]
              (= 2020 (reduce + group))))
    (first)
    (reduce *)))

(defn run []
  (let [input (util/get-input 2020 1)
        numbers (->>
                  (str/split-lines input)
                  (map util/parse-int)
                  (set))]
    [(part-1 numbers)
     (part-2 numbers)]))

File: Advent2020/days/day1.swift
//
//  day1.swift
//  Advent2020
//
//  Created by Omar Estrella on 12/7/20.
//

import Foundation
import Algorithms

private func part1(numbers: [Int]) -> Int {
  return numbers.filter {
    numbers.contains(2020 - $0)
  }.reduce(1, *)
}

private func part2(numbers: [Int]) -> Int {
  return numbers.combinations(ofCount: 3).filter {
    $0.reduce(0, +) == 2020
  }
  .first?
  .reduce(1, *) ?? 0
}

func day1() {
  let numbers = getInput(day: 1).split(separator: "\n").compactMap { Int($0) }
  print("Part 1: \(part1(numbers: numbers))")
  print("Part 2: \(part2(numbers: numbers))")
}
